module/VuFind/src/VuFind/Auth/OpenIDConnect.php

```diff
@@ -0,0 +1,446 @@
+            'email',
+            'cat_id',
+            'cat_username',
+            'cat_password',
+            'college',
+            'major',
+            'home_library',
+        ];
+        $user = $this->getUserTable()->getByUsername($userInfo->sub);
+        $attrMappings = array_filter(
+            $this->getAttributesMappings(),
+            function ($key) use ($availableAttributes) {
+                return in_array($key, $availableAttributes);
+            },
+            ARRAY_FILTER_USE_KEY
+        );
+        $catPassword = null;
+        foreach ($attrMappings as $userAttr => $infoAttr) {
+            $attrValue = $this->getAttributeValue($userInfo, $infoAttr);
+            if (!empty($attrValue)) {
+                if ($userAttr === 'email') {
+                    $user->updateEmail($attrValue);
+                    continue;
+                }
+                if ($userAttr === 'cat_password') {
+                    $catPassword = $attrValue;
+                    continue;
+                }
+                $user->$userAttr = $attrValue;
+            }
+        }
+        if (!empty($user->cat_username)) {
+            $user->saveCredentials($user->cat_username, $catPassword ?? $user->getCatPassword());
+        }
+        $user->save();
+        return $user;
+    }
+
+    /**
+     * Get the URL to establish a session (needed when the internal VuFind login form is inadequate). Returns false
+     * when no session initiator is needed.
+     *
+     * @param string $target Full URL where external authentication method should send user after login (some drivers
+     * may override this).
+     *
+     * @return bool|string
+     */
+    public function getSessionInitiator($target): bool|string
+    {
+        $provider = $this->getProvider();
+        // Adding the auth_method setting makes it possible to handle logins when
+        // using an auth method that proxies others (e.g. ChoiceAuth)
+        $targetUri = $target . (str_contains($target, '?') ? '&' : '?') . 'auth_method=oidc';
+        if (empty($this->session->oidcLastUri) && !empty($target)) {
+            $this->session->oidcLastUri = $targetUri;
+        }
+        $params = [
+            'response_type' => 'code',
+            'redirect_uri' => $targetUri,
+            'client_id' => $this->getConfig()->OpenIDConnect->client_id,
+            'nonce' => $this->session->oidc_nonce,
+            'state' => $this->session->oidc_state,
+            'scope' => 'openid profile email',
+        ];
+        return $provider->authorization_endpoint . '?' . http_build_query($params);
+    }
+
+    /**
+     * Obtain an access token from a code.
+     *
+     * @param string $code Code to look up.
+     *
+     * @return object
+     */
+    protected function getRequestToken(string $code): object
+    {
+        if (isset($this->requestToken)) {
+            return $this->requestToken;
+        }
+        $url = $this->getProvider()->token_endpoint;
```

Comment:
If we're going to assume that certain properties exist in the provider metadata, should getProvider() validate the retrieved data? Or should we add null coalescing here to handle missing fields?